test/dist_autograd_test.py

```diff
@@ -784,6 +782,129 @@ def test_backward_different_dtypes(self):
+        t4 = torch.nn.functional.linear(t2, t3)
+        t5 = torch.nn.functional.linear(t3, t4)
+        return torch.chain_matmul(t1, t2, t3, t4, t5)
+
+    @dist_init
+    def test_backward_complex_python_udf(self):
+        # Run the same code locally and with dist autograd and verify gradients
+        # are same.
+        local_grads = None
+        t1 = torch.rand((3, 3), requires_grad=True)
+        t2 = torch.rand((3, 3), requires_grad=True)
+        for exec_mode in [ExecMode.LOCAL, ExecMode.REMOTE]:
+            with dist_autograd.context() as context_id:
+                ret = self._exec_func(exec_mode, DistAutogradTest._complex_python_udf, t1, t2)
+                loss = ret.sum()
+                local_grads = self._verify_backwards(exec_mode, [loss], context_id, local_grads, t1, t2)
+
+    def _python_udf_with_backward_error(t1, t2):
+        t3 = t1 + t2
+        t4 = SimulateBackwardError.apply(t3)
+        return torch.chain_matmul(t1, t2, t3, t4)
+
+    def _nested_rpc_call_backward_error(t1, t2, dst):
+        t1 = t1 * t2
+        t2 = t1 + t2
+        res = rpc.rpc_sync('worker{}'.format(dst), DistAutogradTest._python_udf_with_backward_error, args=(t1, t2))
+        return torch.chain_matmul(t1, t2, res)
+
+
+    @dist_init
+    def test_backward_python_udf_error(self):
+        t1 = torch.rand((3, 3), requires_grad=True)
+        t2 = torch.rand((3, 3), requires_grad=True)
+        with dist_autograd.context() as context_id:
+            loss = rpc.rpc_sync('worker{}'.format(self._next_rank()), DistAutogradTest._nested_rpc_call_backward_error, args=(t1, t2, self._next_rank()))
+            with self.assertRaisesRegex(RuntimeError, 'Simulate error on backward pass'):
+                dist_autograd.backward([loss.sum()])
+
+
+    _backward_done = False
+
+    def _set_backward_done():
+        DistAutogradTest._backward_done = True
+
+    def _wait_backward_done():
+        while not DistAutogradTest._backward_done:
+            time.sleep(0.1)
+            pass
+
+    @unittest.skipIf(TEST_CONFIG.rpc_backend == RpcBackend.PROCESS_GROUP,
+                     "Skipping this test temporarily since ProcessGroupAgent does not report errors on node failures")
+    @dist_init(clean_shutdown=False)
+    def test_backward_node_failure_python_udf(self):
+        with dist_autograd.context() as context_id:
+            t1 = torch.rand((3, 3), requires_grad=True)
+            t2 = torch.rand((3, 3), requires_grad=True)
+
+            dst = self._next_rank()
+            res = rpc.rpc_sync('worker{}'.format(dst), my_py_nested_call,
+                               args=(t1, t2, dst, self.world_size, 1))
+
+            # Wait for all RPCs to be done.
+            dist.barrier()
+
+            # Kill rank 2 (last hop of nested rpc) and verify rank 0 receives an error.
+            if self.rank == 2:
+                return
+
+            if self.rank == 0:
+                # Wait a bit for rank 2 to die.
+                time.sleep(5)
```

Comment:
This depends on physical time to sync processes, which might potentially make this test flaky. Is there a way to implement the ordering explicitly?